packages/eslint-plugin/src/util/collectUnusedVariables.ts

```diff
@@ -438,6 +448,47 @@ function isExported(variable: TSESLint.Scope.Variable): boolean {
   });
 }
 
+function allVariablesImplicitlyExported(
+  scope: TSESLint.Scope.Scope,
+  isDefinitionFile: boolean,
+): boolean {
+  // TODO: does this also happen in ambient module declarations?
+  if (
+    !isDefinitionFile ||
+    !(
+      scope.type === ScopeType.tsModule ||
+      scope.type === ScopeType.module ||
+      scope.type === ScopeType.global
+    )
+  ) {
+    return false;
+  }
+
+  // TODO: test modules, globals
+  // TODO: look for `export {}`
+
+  function isExportImportEquals(variable: Variable): boolean {
+    for (const def of variable.defs) {
+      if (
+        def.type === TSESLint.Scope.DefinitionType.ImportBinding &&
+        def.node.type === AST_NODE_TYPES.TSImportEqualsDeclaration &&
+        def.node.parent.type === AST_NODE_TYPES.ExportNamedDeclaration
+      ) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  for (const variable of scope.variables) {
+    if (isExported(variable) && !isExportImportEquals(variable)) {
```

Comment:
```ts
declare namespace Foo {
  const x = 1;
  export const y = 3;
}

Foo.x;
Foo.y;
```

The `export` doesn't always restrict other implicit, ambient exports, right?

Reply:
Hm, the rules for this are super annoying. What I have currently may just be true only for declaration files, but "ambient" ones with declare (in ts files only?) may have different rules?

Reply:
It looks like these rules apply to dts files as well
[playground](https://www.typescriptlang.org/play/?filetype=d.ts#code/CYUwxgNghgTiAEA7KBbEBnADlMCBiA9gfAN4BQ88YBi6ALvAB7wC88AjANwXwiOYEYDarQYBPVvADM3AL5kyfAUPh0xmBMzZqNBAGbxCBAHSNuSwQx0IJ29SH2GixsdyA)? Or did I misunderstand you?

Reply:
There are more issues than just this one IIRC; I've been so busy with other stuff that I haven't been able to sit down and actually fix this PR, sorry.

I'm also not 100% sure what works and doesn't work in the playground when it comes to d.ts files; not really something we often play with; I've been meaning to figure out something better for testing this

If it's bugging you all to stay open, I can close it, though if someone else attempts it too I'd love to review it and double check everything (but I don't think anyone's working on it besides me)

Reply:
Nah Josh just pinged me cos I didn't respond :P
No rush at all.